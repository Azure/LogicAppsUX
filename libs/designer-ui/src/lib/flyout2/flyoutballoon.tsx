import { DocumentationLinkItem } from '../documentationlinkitem';
import { calloutContentStyles } from '../fabric';
import type { FlyoutSelectedEventHandler } from './types';
import type { Target } from '@fluentui/react';
import { Callout, DirectionalHint } from '@fluentui/react';

export interface FlyoutBalloonProps {
  documentationLink?: Swagger.ExternalDocumentation;
  flyoutExpanded: boolean;
  target: Target | undefined;
  text: string;
  onClick?: FlyoutSelectedEventHandler;
  onDocumentationLinkClick?(): void;
}

export const FlyoutBalloon: React.FC<FlyoutBalloonProps> = ({
  documentationLink,
  flyoutExpanded,
  target,
  text,
  onClick,
  onDocumentationLinkClick,
}) => {
  if (!flyoutExpanded) {
    return null;
  }

  const handleDismiss = () => {
    onClick?.({ key: '' });
  };

  return (
    <Callout
      beakWidth={8}
      className="msla-flyout-callout"
      directionalHint={DirectionalHint.rightTopEdge}
      gapSpace={0}
      setInitialFocus={true}
      styles={calloutContentStyles}
      target={target}
      onDismiss={handleDismiss}
    >
      <div aria-label={text} data-is-focusable={true} role="dialog" tabIndex={0}>
        {text}
        {documentationLink ? (
          <DocumentationLinkItem
            description={documentationLink.description}
            url={documentationLink.url}
            onClick={onDocumentationLinkClick}
          />
        ) : null}
      </div>
    </Callout>
  );
};
