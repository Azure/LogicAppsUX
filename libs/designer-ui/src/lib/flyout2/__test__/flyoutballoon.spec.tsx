import { DirectionalHint } from '@fluentui/react';
import React from 'react';
import renderer from 'react-test-renderer';
import ShallowRenderer from 'react-test-renderer/shallow';
import { FlyoutBalloon, FlyoutBalloonProps } from '../flyoutballoon';

describe('lib/flyout2/flyoutballoon', () => {
  let minimal: FlyoutBalloonProps;

  beforeEach(() => {
    minimal = {
      flyoutExpanded: false,
      target: undefined,
      text: 'text',
    };
  });

  it('should not render when not expanded', () => {
    const tree = renderer.create(<FlyoutBalloon {...minimal} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  describe('flyoutExpanded', () => {
    let renderer: ShallowRenderer.ShallowRenderer;

    beforeEach(() => {
      renderer = ShallowRenderer.createRenderer();
    });

    afterEach(() => {
      renderer.unmount();
    });

    it('should render when expanded', () => {
      renderer.render(<FlyoutBalloon {...minimal} flyoutExpanded={true} />);

      const balloon = renderer.getRenderOutput();
      expect(balloon.props).toEqual(
        expect.objectContaining({
          beakWidth: 8,
          className: 'msla-flyout-callout',
          directionalHint: DirectionalHint.rightTopEdge,
          gapSpace: 0,
          setInitialFocus: true,
        })
      );

      const dialog = React.Children.only(balloon.props.children);
      expect(dialog.props).toEqual(
        expect.objectContaining({
          'aria-label': minimal.text,
          'data-is-focusable': true,
          role: 'dialog',
          tabIndex: 0,
        })
      );

      const [text]: any[] = React.Children.toArray(dialog.props.children);
      expect(text).toBe(minimal.text);
    });

    it('should render with a documentation link', () => {
      const documentationLink = {
        url: 'https://aka.ms/logicapps-chunk',
      };
      renderer.render(<FlyoutBalloon {...minimal} documentationLink={documentationLink} flyoutExpanded={true} />);

      const balloon = renderer.getRenderOutput();
      const dialog = React.Children.only(balloon.props.children);
      const [, documentationLinkItem]: any[] = React.Children.toArray(dialog.props.children);
      expect(documentationLinkItem.props).toEqual({
        url: documentationLink.url,
      });
    });
  });
});
