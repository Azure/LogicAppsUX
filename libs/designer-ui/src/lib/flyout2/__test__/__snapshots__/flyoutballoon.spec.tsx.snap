// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`lib/flyout2/flyoutballoon > should not render when not expanded 1`] = `null`;
